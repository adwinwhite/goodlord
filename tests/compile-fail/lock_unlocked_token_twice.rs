use deadlock_proof::LockAfter;
use deadlock_proof::impl_lock_order;
use deadlock_proof::LockToken;
use deadlock_proof::GoodLock;
use deadlock_proof::Unlocked;


struct A;
struct B;

impl_lock_order!(Unlocked => A);
impl_lock_order!(A => B);

fn main() {
    let mut unlocked_token = LockToken::UNLOCKED;
    let lock_a: GoodLock<A, _> = GoodLock::new(0_i32);
    let lock_b: GoodLock<B, _> = GoodLock::new(0_i32);
    let (a, _) = lock_a.lock(&mut unlocked_token);
    let (b, _) = lock_b.lock(&mut unlocked_token);
    println!("{}, {}", a, b);
}
