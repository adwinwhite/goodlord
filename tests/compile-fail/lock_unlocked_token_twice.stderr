error[E0499]: cannot borrow `unlocked_token` as mutable more than once at a time
  --> tests/compile-fail/lock_unlocked_token_twice.rs:19:30
   |
18 |     let (a, _) = lock_a.lock(&mut unlocked_token);
   |                              ------------------- first mutable borrow occurs here
19 |     let (b, _) = lock_b.lock(&mut unlocked_token);
   |                              ^^^^^^^^^^^^^^^^^^^ second mutable borrow occurs here
20 |     println!("{}, {}", a, b);
   |                        - first borrow later used here
